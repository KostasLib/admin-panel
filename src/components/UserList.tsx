import React from 'react';
import '../ListTile.scss';

interface IUser {
  // Properties useful to the list
  id: string;
  name: string;
  email?: string;
  photo?: string;

  // Anything else that is not of interest to the list
  [key: string]: any;
}

interface IProps {
  users: IUser[];
  onSelectUser: (id: string) => () => void;
  selected?: string;
}

export const UserList: React.FC<IProps> = props => {
  const { users, selected, onSelectUser } = props;

  return (
    <ul>
      {users.map(user => (
        <li
          key={user.id}
          className={`list-tile ${
            selected === user.id ? 'list-tile--selected' : 'list-tile--with-hover'
          }`}
          onClick={onSelectUser(user.id)}
        >
          <div className="list-tile__avatar">
            <img
              className="list-tile__avatar__image"
              src={user.photo}
              alt={`${user.name}'s avatar`}
            />
          </div>
          <div className="list-tile__details">{user.name}</div>
        </li>
      ))}
    </ul>
  );
};
