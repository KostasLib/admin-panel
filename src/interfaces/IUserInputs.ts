export interface IUSerInputs {
  name: string;
  email: string;
  phone: string;
  address: string;
  company: string;
}
