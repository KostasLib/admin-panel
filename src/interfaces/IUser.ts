import { IUSerInputs } from './IUserInputs';

export interface IUser extends IUSerInputs {
  id: string;
  photo?: string;
}
