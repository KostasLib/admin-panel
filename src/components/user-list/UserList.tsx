import React from 'react';
import { IUser } from '../../interfaces/IUser';
import { ListTile } from './ListTile';

interface IProps {
  users: IUser[];
  onSelectUser: (id: string) => () => void;
  selected?: string;
}

/** A stateless component that maps through and array of users an renders a list tile for each user */
export const UserList: React.FC<IProps> = props => {
  const { users, selected, onSelectUser } = props;

  return (
    <ul>
      {users.map(user => (
        <ListTile
          user={user}
          selected={selected === user.id}
          title={user.name}
          onClick={onSelectUser(user.id)}
          key={user.id}
        />
      ))}
    </ul>
  );
};
