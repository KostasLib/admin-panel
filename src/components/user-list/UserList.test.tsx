import React from 'react';
import { shallow } from 'enzyme';
import { UserList } from './UserList';
import { IUser } from '../../interfaces/IUser';

describe('The <UserList /> component', () => {
  const users: IUser[] = [
    {
      id: '5c093af1c6ee9117a581c7d6',
      photo: 'https://randomuser.me/api/portraits/men/40.jpg',
      name: 'Bates Washington',
      company: 'ZOLAREX',
      email: 'bates.washington@zolarex.io',
      phone: '+1 (915) 447-2207',
      address: '958 Brevoort Place, Ona, Maine, 2433',
    },
    {
      id: '5c093af1aeca1bb00607fb2a',
      photo: 'https://randomuser.me/api/portraits/men/1.jpg',
      name: 'Mollie Oneill',
      company: 'VIAGRAND',
      email: 'mollie.oneill@viagrand.biz',
      phone: '+1 (852) 535-3880',
      address: '120 Cedar Street, Mansfield, Kentucky, 8890',
    },
  ];
  const mockClickUser = jest.fn(() => null);
  const mockSelectUser = jest.fn((id: string) => mockClickUser);

  // Assert that an event handler is generated for each list tile.
  afterEach(() => expect(mockSelectUser).toHaveBeenCalledWith(users[0].id));

  // Snapshots
  it('maps through an array of users and renders a <ListTile /> for each user', () => {
    const component = shallow(<UserList users={users} onSelectUser={mockSelectUser} />);

    expect(component).toMatchSnapshot();
  });

  it('renders an empty list if the users array is empty', () => {
    const component = shallow(<UserList users={[]} onSelectUser={mockSelectUser} />);

    expect(component).toMatchSnapshot();
  });

  // Event
  it('passes an event handler on each child that can be called on click', () => {
    const component = shallow(<UserList users={users} onSelectUser={mockSelectUser} />);
    const node = component.findWhere(node => node.key() === users[0].id);

    // Assert that the event handler is called properly.
    expect(mockClickUser).not.toHaveBeenCalled();
    node.simulate('click');
    expect(mockClickUser).toHaveBeenCalled();
  });
});
