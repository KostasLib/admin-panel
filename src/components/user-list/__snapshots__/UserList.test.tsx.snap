// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <UserList /> component maps through an array of users and renders a <ListTile /> for each user 1`] = `
<ul>
  <ListTile
    key="5c093af1c6ee9117a581c7d6"
    onClick={[MockFunction]}
    selected={false}
    title="Bates Washington"
    user={
      Object {
        "address": "958 Brevoort Place, Ona, Maine, 2433",
        "company": "ZOLAREX",
        "email": "bates.washington@zolarex.io",
        "id": "5c093af1c6ee9117a581c7d6",
        "name": "Bates Washington",
        "phone": "+1 (915) 447-2207",
        "photo": "https://randomuser.me/api/portraits/men/40.jpg",
      }
    }
  />
  <ListTile
    key="5c093af1aeca1bb00607fb2a"
    onClick={[MockFunction]}
    selected={false}
    title="Mollie Oneill"
    user={
      Object {
        "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
        "company": "VIAGRAND",
        "email": "mollie.oneill@viagrand.biz",
        "id": "5c093af1aeca1bb00607fb2a",
        "name": "Mollie Oneill",
        "phone": "+1 (852) 535-3880",
        "photo": "https://randomuser.me/api/portraits/men/1.jpg",
      }
    }
  />
</ul>
`;

exports[`The <UserList /> component renders an empty list if the users array is empty 1`] = `<ul />`;
